src/main/java/io/aiven/kafka/connect/http/sender/OAuth2HttpSender.java

```diff
@@ -19,82 +19,98 @@
+        final int retriesNumber
+    ) {
+        final HttpResponseHandler handler = (response, retriesRemain) -> {
+            if (response.statusCode() == 401 && retriesRemain > 0) { // access denied or refresh of a token is needed
+                ((OAuth2AuthHttpRequestBuilder) this.httpRequestBuilder).renewAccessToken(requestBuilderWithPayload);
+                this.sendWithRetries(requestBuilderWithPayload, originHttpResponseHandler, retriesRemain - 1);
             } else {
-                originHttpResponseHandler.onResponse(response);
+                originHttpResponseHandler.onResponse(response, retriesRemain);
             }
         };
-        return super.sendWithRetries(accessTokenAwareRequestBuilder, handler);
+        return super.sendWithRetries(requestBuilderWithPayload, handler, retriesNumber);
     }
 
-    private HttpRequest.Builder loadOrRefreshAccessToken(final HttpRequest.Builder requestBuilder) {
-        if (accessTokenAuthHeader == null) {
-            logger.info("Configure OAuth2 for URI: {} and Client ID: {}",
-                    config.oauth2AccessTokenUri(), config.oauth2ClientId());
+    static class OAuth2AuthHttpRequestBuilder extends DefaultHttpRequestBuilder {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(OAuth2AuthHttpRequestBuilder.class);
+        private final ObjectMapper objectMapper = new ObjectMapper();
+        private String accessTokenAuthHeader;
+        private final AccessTokenHttpSender accessTokenHttpSender;
+        private final HttpSinkConfig config;
+
+        OAuth2AuthHttpRequestBuilder(final HttpSinkConfig config, final AccessTokenHttpSender accessTokenHttpSender) {
+            this.config = config;
+            this.accessTokenHttpSender = accessTokenHttpSender;
+        }
+
+        @Override
+        public Builder build(final HttpSinkConfig config) {
+            return super
+                .build(config)
+                .header(HEADER_AUTHORIZATION, getAccessToken());
+        }
+
+        void renewAccessToken(final HttpRequest.Builder requestBuilderWithPayload) {
+            this.accessTokenAuthHeader = null;
+            requestBuilderWithPayload.setHeader(HttpRequestBuilder.HEADER_AUTHORIZATION, this.getAccessToken());
+        }
+
+        private String getAccessToken() {
+            if (this.accessTokenAuthHeader != null) {
+                return this.accessTokenAuthHeader;
+            }
+            LOGGER.info("Configure OAuth2 for URI: {} and Client ID: {}", config.oauth2AccessTokenUri(),
+                config.oauth2ClientId());
             try {
-                final var response =
-                        super.sendWithRetries(
-                                new AccessTokenHttpRequestBuilder().build(config),
-                                HttpResponseHandler.ON_HTTP_ERROR_RESPONSE_HANDLER
-                        );
+                final var response = accessTokenHttpSender.call();
```

Comment:
The original source was hard to understand when we had to follow the access token retrieval. I tried to make it clearer with one method  and 2 references:
First when the Builder is built:
```        
@Override
        public Builder build(final HttpSinkConfig config) {
            return super
                .build(config)
                // We need to retrieve an access token first
                .header(HEADER_AUTHORIZATION, getAccessToken());
        }
```
This is a normal process when we need to init the session, and we don't need to fail before trying to retrieve one.

Then, whenever a response is not successful in the super class:
````
 try {
                    final var response =
                            httpClient.send(requestBuilderWithPayload.build(), HttpResponse.BodyHandlers.ofString());
                    log.debug("Server replied with status code {} and body {}", response.statusCode(), response.body());
                    // Handle the response
                    httpResponseHandler.onResponse(response, remainingRetries);
                    return response;
                }
````
We let the sub class handle the error. If possible a `renewAccessToken` is called, otherwise we let the interface throw an error:
````
  final HttpResponseHandler handler = (response, remainingRetries) -> {
            // If the response has a 401 error and we have retries left, we attempt to renew the session
            if (response.statusCode() == 401 && remainingRetries > 0) {
                // Update the request builder with the new access token
                ((OAuth2AuthHttpRequestBuilder) this.httpRequestBuilder).renewAccessToken(requestBuilder);
                // Retry the call and decrease the retries counter to avoid looping on token renewal
                this.sendWithRetries(requestBuilder, originHttpResponseHandler, remainingRetries - 1);
            } else {
                originHttpResponseHandler.onResponse(response, remainingRetries);
            }
        };
````

That way we clearly see where a token is requested and when it's done.

Reply:
I'm not sure why you're telling me the compatibility is broken here? I kept the initial behavior which was to retry request if the token was expired. Tests were made to verify this behavior:

- `OAuth2HttpSenderTest1#throwsConnectExceptionForUnauthorizedToken`
- `OAuth2HttpSenderTest#refreshAccessToken`

Reply:
Totally agree about the refactoring. The separation of getting first access token, and the refresh when needed is excellent.

About the behavior change, you're right. I was wrongly reading the `send` as individual call but -- as you pointed in a separate comment -- it's actually retrying internally -- my apologies. Will mark this one as solved then. Thanks for explaining!